File: trees_and_graphs/01_matrix.py
from collections import deque

def updateMatrix(mat: list[list[int]]) -> list[list[int]]:
    def valid(row, col):
        return 0 <= row < m and 0 <= col < n and mat[row][col] == 1

    # Can create copy of mat to avoid modifying the input
    m = len(mat)
    n = len(mat[0])

    queue = deque();
    seen = set();

    for row in range(m):
        for col in range(n):
            if mat[row][col] == 0:
                queue.append((row, col, 1))
                seen.add((row, col))
    
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    while queue:
        row, col, steps = queue.popleft()

        for dcol, drow in directions:
            next_row, next_col = row + drow, col + dcol

            if valid(next_row, next_col) and (next_row, next_col) not in seen:
                queue.append((next_row, next_col, steps + 1))
                seen.add((next_row, next_col))
                mat[next_row][next_col] = steps
    
    return mat

File: trees_and_graphs/01_matrix.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function(mat)
{
    const valid = (row, col) => {
        return 0 <= row && row < m && 0 <= col && col < n && mat[row][col] == 1;
    }

    // Can create a copy of mat to avoid modifying the input
    const m = mat.length;
    const n = mat[0].length;

    let queue = [];
    let seen = [];
    for (let i = 0; i < m; i++)
    {
        seen.push(new Array(n).fill(false));
    }

    for (let row = 0; row < m; row++)
    {
        for (let col = 0; col < n; col++)
        {
            if (mat[row][col] == 0)
            {
                queue.push([row, col]);
                seen[row][col] = true;
            }
        }
    }

    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    let steps = 0;

    while (queue.length)
    {
        let currentLength = queue.length;
        let nextQueue = [];
        steps++;

        for (let i = 0; i < currentLength; i++)
        {
            const [row, col] = queue[i];
            for (const [dcol, drow] of directions)
            {
                const nextRow = row + drow;
                const nextCol = col + dcol;

                if (valid(nextRow, nextCol) && !seen[nextRow][nextCol])
                {
                    seen[nextRow][nextCol] = true;
                    nextQueue.push([nextRow, nextCol]);
                    mat[nextRow][nextCol] = steps;
                }
            }
        }

        queue = nextQueue;
    }

    return mat;
};